### ISSUE TITLE
[2015 ICCV Oral] Fast R-CNN

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#29]
기존의 R-CNN, SPPnet이 가지고 있던 단점인 multi-stage pipeline, expensive computing, slow speed를 극복한 논문

즉 다음과 같은 장점을 가지고 있음
1. Single-stage, using multi-task loss
2. Training can update all network layers
3. No disk storage is required for feature caching

Fast R-CNN, RoI pooling, 실험만 설명

### 1.1. How it works
![image](https://user-images.githubusercontent.com/49643709/123536548-d7f8af00-d765-11eb-8286-0463f98cc79b.png)

1. Input으로 image와 region proposal이 주어짐 (selective search를 통해 찾음)
2. Input이 주어졌을 때 image를 CNN에 넣음
3. Feature tensor에서 region에 해당하는 부분을 추출 (SPPnet과 동일)
4. RoI pooling을 통해 fixed-size feature tensor로 만듬
5. Fixed-size feature tensor는 fc layers를 거친 후 classification, bounding box regression fc layers를 병렬적으로 넣음
6. Output으로 classification에 수행될 softmax, 수정된 region proposals를 얻음
    6.1. Train : backprop을 통해 네트워크 학습
    6.2. Inference : confidence 값을 기준으로 filter + NMS를 통해 bounding box prediction 수행 (R-CNN과 동일)

### 1.2. RoI pooling

Input이 어떻든 간에 정해진 size를 얻기 위한 방법으로, pytorch 기준으로 adaptive max pooling임

SPP에서 소개된 spatial pyramid pooling의 일부임 (special case of the spatial pyramid pooling layer, which is one pyramid level)

CNN model을 VGG 16으로 사용했을 때를 기준으로 7 * 7 RoI pooling 사용

### 1.3. CNN model (VGG 16 기준)

- Last max pooling layer를 RoI pooling layer로 교체
- Last fully connected layer and softmax를 2 sibling layers로 교체 (category fc, category-specific bounding box fc)

### 1.4. Multi-task loss
![image](https://user-images.githubusercontent.com/49643709/123537277-8e11c800-d769-11eb-9a5a-027709f04bde.png)
![image](https://user-images.githubusercontent.com/49643709/123537282-92d67c00-d769-11eb-990f-0b5497428c0e.png)

- Loss는 classification loss, localization loss 2 파트로 나뉨
    - Classification loss : cross-entropy
    - Localization loss : smooth L1 loss
- lambda = 1로 설정

### 1.5. Sampling

총 2장을 mini-batch에 넣으며, 1장의 image에 대하여 64 RoI를 sampling하여 학습
- RoI 25% : positive (g.t. IoU와의 overlap이 0.5 이상인 경우)
- RoI 75% : negative (g.t. IoU와의 overlap이 0.1 이상 0.5 미만인 경우)
- No data augmentation except random horizontal flip

### 2.1. Scale invariance
![image](https://user-images.githubusercontent.com/49643709/123537676-ab479600-d76b-11eb-8c63-8365d0047e3d.png)

Scale-invariant object detection을 위해 기존에 2가지 방법이 있었음
1. Brute-force learning (single scale)
2. Image pyramids (multi-scale)

Single scale experiment는 600 pixel, multi-scale experiment는 [480, 576, 688, 864, 1200] 5가지 scale에서 학습

Multi-scale이 당연히 더 성능면에서 우수하지만, single scale도 비슷하게 잘 작동함
(Tradeoff between speed and accuracy)

### 2.2. Which layers to fine-tune
![image](https://user-images.githubusercontent.com/49643709/123537524-d7164c00-d76a-11eb-97f8-858ebfc5bb10.png)

- 다른 conv layer의 depth에 따라 freeze, train을 조절하여 실험 진행
    - 결과를 보면 알 수 있듯이, conv layer도 freeze를 풀고 학습하면 더욱 성능이 향상됨
- Conv1 layer를 학습하도록 변경해도 conv2_1부터 학습한 모델과 mAP의 변화는 없었음
    - 이로부터 conv1 layer는 generic하고 task independent함을 확인할 수 있음
    - 즉 모든 conv layers가 fine-tuned될 필요는 없음

### 2.3. Multi-task training
![image](https://user-images.githubusercontent.com/49643709/123537629-6faccc00-d76b-11eb-99d7-8deee6a81a59.png)

- Fast R-CNN은 multi-task loss로 학습하는데, 이가 stage-wise training보다 더 우수한지 실험
- 실험 결과 multi-task training이 stage-wise training보다 뛰어남

### 2.4. More training data

- 기존의 DPM과 같은 모델들은 data가 많더라도 성능이 올라가지 않음
- 하지만 Faster R-CNN 같은 경우, 데이터가 많을 수록 성능이 향상됨

### 2.5. SVM outperform softmax
![image](https://user-images.githubusercontent.com/49643709/123537842-93bcdd00-d76c-11eb-97ae-b38b83f60a36.png)

- R-CNN에서는 softmax를 사용한 것이 SVM을 사용한 것보다 성능이 안좋았음
- Faster R-CNN에서는 softmax를 사용한 것이 성능이 더 좋음

### 2.6. More proposals and other proposals
![image](https://user-images.githubusercontent.com/49643709/123537895-e0081d00-d76c-11eb-87d0-d419abb5cd7a.png)

Object proposals로 2가지 종류가 있음
1. Sparse set : selective search
2. Dense set : DPM

- 먼저 selective search를 통해 찾은 proposal을 늘리면 어떻게 되는지 실험
    - 증가하다 감소함
- Selective search 대신 DPM proposal을 늘림
    - 감소
- Only DPM
    - Selective search 대비 성능 감소
- 결국 selective search로 찾는 것이 가장 효과적이었으며, proposal을 더 늘릴 필요가 없음
